File: static/js/subscription/add_payment_method.js
document.addEventListener('DOMContentLoaded', function () {
    const form = document.getElementById('addPaymentMethodForm');
    const saveButton = document.getElementById('savePaymentMethodBtn');
    const inputs = form.querySelectorAll('input[required]');

    // Функция проверки номера карты по алгоритму Луна
    function isValidCardNumber(cardNumber) {
        // Удаляем пробелы и проверяем, что это только цифры
        const cleanNumber = cardNumber.replace(/\D/g, '');
        if (!/^\d{13,19}$/.test(cleanNumber)) {
            return false;
        }

        let sum = 0;
        let isEven = false;
        for (let i = cleanNumber.length - 1; i >= 0; i--) {
            let digit = parseInt(cleanNumber[i]);
            if (isEven) {
                digit *= 2;
                if (digit > 9) digit -= 9;
            }
            sum += digit;
            isEven = !isEven;
        }
        return sum % 10 === 0;
    }

    // Проверка срока действия (MM/YY, не истёк)
    function isValidExpiryDate(expiry) {
        const match = expiry.match(/^(\d{2})\/(\d{2})$/);
        if (!match) return false;

        const month = parseInt(match[1]);
        const year = parseInt(match[2]) + 2000; // Предполагаем 20XX
        const now = new Date();
        const currentYear = now.getFullYear();
        const currentMonth = now.getMonth() + 1;

        return month >= 1 && month <= 12 && year >= currentYear &&
               (year > currentYear || (year === currentYear && month >= currentMonth));
    }

    // Проверка CVV (3 или 4 цифры)
    function isValidCVV(cvv, cardNumber) {
        const cleanNumber = cardNumber.replace(/\D/g, '');
        const isAmEx = /^3[47]/.test(cleanNumber);
        const cvvPattern = isAmEx ? /^\d{4}$/ : /^\d{3}$/;
        return cvvPattern.test(cvv);
    }

    saveButton.addEventListener('click', function () {
        let isValid = true;
        const cardNumberInput = document.getElementById('cardNumber');
        const expiryDateInput = document.getElementById('expiryDate');
        const cvvInput = document.getElementById('cvv');
        const cardholderNameInput = document.getElementById('cardholderName');

        // Проверка обязательных полей
        inputs.forEach(input => {
            if (!input.value.trim()) {
                input.classList.add('is-invalid');
                isValid = false;
            } else {
                input.classList.remove('is-invalid');
            }
        });

        // Проверка номера карты
        if (cardNumberInput.value.trim()) {
            if (!isValidCardNumber(cardNumberInput.value)) {
                cardNumberInput.classList.add('is-invalid');
                isValid = false;
            }
        }

        // Проверка срока действия
        if (expiryDateInput.value.trim()) {
            if (!isValidExpiryDate(expiryDateInput.value)) {
                expiryDateInput.classList.add('is-invalid');
                isValid = false;
            }
        }

        // Проверка CVV
        if (cvvInput.value.trim()) {
            if (!isValidCVV(cvvInput.value, cardNumberInput.value)) {
                cvvInput.classList.add('is-invalid');
                isValid = false;
            }
        }

        // Если все проверки пройдены, отправляем форму
        if (isValid) {
            const formData = new FormData(form);

            fetch('/add-payment-method/', {
                method: 'POST',
                body: formData,
                headers: {
                    'X-CSRFToken': getCookie('csrftoken')
                }
            })
            .then(response => response.json())
            .then(data => {
                if (data.status === 'success') {
                    const modal = bootstrap.Modal.getInstance(document.getElementById('addPaymentMethodModal'));
                    modal.hide();
                    alert(data.message);
                } else {
                    alert(window.translations.error_processing_form + ': ' + (data.message || window.translations.unknown_error));
                }
            })
            .catch(error => {
                console.error('Error:', error);
                alert(window.translations.error_sending_form);
            });
        }
    });

    // Удаление класса is-invalid при вводе данных
    inputs.forEach(input => {
        input.addEventListener('input', function () {
            if (input.value.trim()) {
                input.classList.remove('is-invalid');
            }
        });
    });

    // Функция для получения CSRF-токена
    function getCookie(name) {
        let cookieValue = null;
        if (document.cookie && document.cookie !== '') {
            const cookies = document.cookie.split(';');
            for (let i = 0; i < cookies.length; i++) {
                const cookie = cookies[i].trim();
                if (cookie.substring(0, name.length + 1) === (name + '=')) {
                    cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
                    break;
                }
            }
        }
        return cookieValue;
    }
});

File: apps/subscribe/add_payment_method.py
from django.http import JsonResponse
from project.language import translate, language
from django.views.decorators.http import require_POST
import logging
import re
from datetime import datetime

# Настройка логирования для вывода в консоль
logger = logging.getLogger(__name__)

@require_POST
def add_payment_method(request):
    """
    Обработчик для добавления способа оплаты.
    Проверяет валидность данных карты и выводит их в консоль.
    """
    # Извлечение данных из POST-запроса
    lang = language(request)
    card_number = request.POST.get('cardNumber')
    expiry_date = request.POST.get('expiryDate')
    cvv = request.POST.get('cvv')
    cardholder_name = request.POST.get('cardholderName')
    make_default = request.POST.get('makeDefaultCard') == 'on'

    # Проверка обязательных полей
    if not all([card_number, expiry_date, cvv, cardholder_name]):
        return JsonResponse({
            'status': 'error',
            'message': translate("Все поля обязательны для заполнения.", lang)
        }, status=400)

    # Проверка номера карты (алгоритм Луна)
    def is_valid_card_number(card_number):
        clean_number = re.sub(r'\D', '', card_number)
        if not re.match(r'^\d{13,19}$', clean_number):
            return False
        sum = 0
        is_even = False
        for i in range(len(clean_number) - 1, -1, -1):
            digit = int(clean_number[i])
            if is_even:
                digit *= 2
                if digit > 9:
                    digit -= 9
            sum += digit
            is_even = not is_even
        return sum % 10 == 0

    if not is_valid_card_number(card_number):
        return JsonResponse({
            'status': 'error',
            'message': translate("Недействительный номер карты", lang)
        }, status=400)

    # Проверка срока действия
    def is_valid_expiry_date(expiry):
        match = re.match(r'^(\d{2})\/(\d{2})$', expiry)
        if not match:
            return False
        month, year = map(int, match.groups())
        year += 2000
        now = datetime.now()
        return 1 <= month <= 12 and year >= now.year and \
               (year > now.year or (year == now.year and month >= now.month))

    if not is_valid_expiry_date(expiry_date):
        return JsonResponse({
            'status': 'error',
            'message': translate("Недействительный или истёкший срок действия.", lang)
        }, status=400)

    # Проверка CVV
    def is_valid_cvv(cvv, card_number):
        clean_number = re.sub(r'\D', '', card_number)
        is_amex = re.match(r'^3[47]', clean_number)
        cvv_pattern = r'^\d{4}$' if is_amex else r'^\d{3}$'
        return bool(re.match(cvv_pattern, cvv))

    if not is_valid_cvv(cvv, card_number):
        return JsonResponse({
            'status': 'error',
            'message': translate("Недействительный CVV", lang)
        }, status=400)

    # Вывод данных в консоль
    logger.info("Получены данные формы добавления способа оплаты:")
    logger.info(f"Номер карты: {card_number}")
    logger.info(f"Срок действия: {expiry_date}")
    logger.info(f"CVV: {cvv}")
    logger.info(f"Имя держателя карты: {cardholder_name}")
    logger.info(f"Сделать основным способом оплаты: {make_default}")

    # Возвращаем успешный JSON-ответ
    return JsonResponse({
        'status': 'success',
        'message': translate("Карта успешно добавлена", lang)
    })
